# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/mayurkharche01/Internship-starter-flyrank/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*



I am going with **Lane 2 — Refresh / Content Opportunity Scoring**. I want to find which content items may be worth reviewing first. The dataset has useful information about traffic, clicks, engagement, content age, and recent performance, so this lane seems practical to explore further.

**CODE CELL**

In [6]:
import pandas as pd

DATA_PATH = "../../data/raw/content_refresh_anonymized.csv"

df = pd.read_csv(DATA_PATH)

print(f"Rows: {df.shape[0]:,}")
print(f"Columns: {df.shape[1]:,}")
print(f"Unique clients: {df['client_id'].nunique():,}")
print(f"Unique content items: {df['content_id'].nunique():,}")

Rows: 30,000
Columns: 44
Unique clients: 32
Unique content items: 30,000


## 2. The question: decision, action, cost of a wrong call

The main question is: **Which content items should be reviewed first?**

The SEO/content team could use the ranking to decide which pages to check and whether they need a refresh, expansion, protection, pruning, or monitoring.

A wrong recommendation could waste the team's review time. Missing an important page could also mean missing a useful content opportunity.

**Unit of analysis:** one content item.

**CODE CELL**

In [7]:
trend_counts = df["trend_direction"].value_counts(dropna=False)
trend_pct = (
    df["trend_direction"]
    .value_counts(normalize=True, dropna=False)
    .mul(100)
    .round(1)
)

print("Trend direction counts:")
print(trend_counts)

print("\nTrend direction percentages:")
print(trend_pct)

Trend direction counts:
trend_direction
down      16262
stable     5962
up         4388
new        2236
flat       1152
Name: count, dtype: int64

Trend direction percentages:
trend_direction
down      54.2
stable    19.9
up        14.6
new        7.5
flat       3.8
Name: proportion, dtype: float64


## 3. Quick look at the data (2-3 real numbers)

The starter dataset has **30,000 content items from 32 clients**. I found that **16,262 items (54.2%) have a downward trend** in the current data. This gives enough variation to make content prioritization worth exploring.

**CODE CELL**

In [8]:
# Real numbers used in this section

total_items = df["content_id"].nunique()
total_clients = df["client_id"].nunique()

down_count = (df["trend_direction"] == "down").sum()
down_pct = (down_count / len(df)) * 100

print(f"Unique content items: {total_items:,}")
print(f"Unique clients: {total_clients:,}")
print(f"Items with trend_direction='down': {down_count:,} ({down_pct:.1f}%)")

Unique content items: 30,000
Unique clients: 32
Items with trend_direction='down': 16,262 (54.2%)


It's Optional

In [9]:
last30_impressions = df["impressions_last_30d"].sum()
prev30_impressions = df["impressions_prev_30d"].sum()

last30_clicks = df["clicks_last_30d"].sum()
prev30_clicks = df["clicks_prev_30d"].sum()

print(f"Last-30-day impressions: {last30_impressions:,}")
print(f"Previous-30-day impressions: {prev30_impressions:,}")
print(f"Last-30-day clicks: {last30_clicks:,}")
print(f"Previous-30-day clicks: {prev30_clicks:,}")

Last-30-day impressions: 42,871,762
Previous-30-day impressions: 53,492,355
Last-30-day clicks: 148,016
Previous-30-day clicks: 163,053


## 4. Careful words: what I can and can't claim

From the current data, I can describe patterns and use them for decision support.

I cannot say that a page will definitely decline, that refreshing it will improve performance, or that ML will be better than a simple rule. Those things will need to be tested later.

In [10]:
# Data-quality check for a known starter-data gotcha:
# avg_position = 0 means no position data, not rank zero.

zero_position = (df["avg_position"] == 0).sum()

print(f"Rows with avg_position = 0: {zero_position:,}")

Rows with avg_position = 0: 1,205


## Self-check

- [x] Selected a provisional lane
- [x] Defined the question, decision, action, and cost
- [x] Used real numbers from the starter data
- [x] Defined the unit of analysis
- [x] Kept the claims careful
- [ ] Committed the notebook to the repo
- [ ] Submitted the repo URL